lib/Dialect/SV/SVOps.cpp

```diff
@@ -853,6 +853,58 @@ static LogicalResult verifyAliasOp(AliasOp op) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// PAssignOp
+//===----------------------------------------------------------------------===//
+
+// reg s <= cond ? val : s simplification.
+// Don't assign a register's value to itself, conditionally assign the new value
+// instead.
+LogicalResult PAssignOp::canonicalize(PAssignOp op, PatternRewriter &rewriter) {
+  auto mux = op.src().getDefiningOp<comb::MuxOp>();
+  if (!mux)
+    return failure();
+
+  auto reg = dyn_cast<sv::RegOp>(op.dest().getDefiningOp());
+  if (!reg)
+    return failure();
+
+  bool trueBranch; // did we find the register on the true branch?
+  auto tvread = mux.trueValue().getDefiningOp<sv::ReadInOutOp>();
+  auto fvread = mux.falseValue().getDefiningOp<sv::ReadInOutOp>();
+  if (tvread && reg == tvread.input().getDefiningOp<sv::RegOp>())
+    trueBranch = true;
+  else if (fvread && reg == fvread.input().getDefiningOp<sv::RegOp>())
+    trueBranch = false;
+  else
+    return failure();
+
+  // Check that this is the only write of the register
+  for (auto &use : reg->getUses()) {
+    if (isa<ReadInOutOp>(use.getOwner()))
+      continue;
+    if (use.getOwner() == op)
+      continue;
+    return failure();
+  }
+
+  if (trueBranch) {
+    rewriter.create<sv::IfOp>(
+        mux.getLoc(), mux.cond(), [&]() {},
+        [&]() {
+          rewriter.create<PAssignOp>(op.getLoc(), reg, mux.falseValue());
+        });
```

Comment:
Instead of creating the code in the else block, I'd recommend emitting a `xor cond, 1` and emitting it in the true block.  This generates more canonical IR rather than waiting for another canonicalization pattern to flip the if.